scio-core/src/main/scala/com/spotify/scio/io/ZstdDictIO.scala

```diff
@@ -0,0 +1,139 @@
+    val trainingBytesTargetActual: Int = Option(trainingBytesTarget).getOrElse {
+      val computed =
+        Try(Math.multiplyExact(zstdDictSizeBytes, 100)).toOption.getOrElse(Int.MaxValue)
+      logger.info(s"No trainingBytesTarget passed, using ${computed} bytes")
+      computed
+    }
+    if (trainingBytesTargetActual <= 0) {
+      throw new IllegalArgumentException(
+        s"trainingBytesTarget must be positive, found $trainingBytesTargetActual"
+      )
+    }
+
+    val beamCoder = CoderMaterializer.beam(data.context, data.coder)
+    def toBytes(v: T): Array[Byte] = CoderUtils.encodeToByteArray(beamCoder, v)
+
+    data
+      .transform("Create Zstd Dictionary") { scoll =>
+        // estimate the sample rate we need by examining numElementsForSizeEstimation elements
+        val streamsCntSI = scoll.count.asSingletonSideInput(0L)
+        val sampleRateSI = scoll
+          .take(numElementsForSizeEstimation)
+          .map(v => toBytes(v).length)
+          .sum
+          .withSideInputs(streamsCntSI)
+          .map { case (totalSize, ctx) =>
+            val avgSize = totalSize / numElementsForSizeEstimation
+            val targetNumElements = trainingBytesTargetActual / avgSize
+            val sampleRate = targetNumElements.toDouble / ctx(streamsCntSI)
+            logger.info(s"Computed sample rate for Zstd dictionary: ${sampleRate}")
+            sampleRate
+          }
+          .toSCollection
+          .asSingletonSideInput
+
+        scoll
+          .withSideInputs(sampleRateSI)
+          .flatMap {
+            case (s, ctx) if new Random().nextDouble() <= ctx(sampleRateSI) =>
+              Some(toBytes(s))
+            case _ => None
+          }
+          .toSCollection
+          .keyBy(_ => ())
+          .groupByKey
+          .map { case (_, elements) =>
+            val zstdSampleSize = {
+              val sum = elements.map(_.length.toLong).sum
+              if (sum > Int.MaxValue.toLong) Int.MaxValue else sum.toInt
+            }
+            logger.info(s"Training set size for for Zstd dictionary: ${zstdSampleSize}")
+            val trainer = new ZstdDictTrainer(zstdSampleSize, zstdDictSizeBytes)
+            elements.foreach(trainer.addSample)
+            trainer.trainSamples()
+          }
+      }
+      .withName("Save Zstd Dictionary")
+      .saveAsBinaryFile(path)
+      .underlying
+  }
+
+  override def tap(read: Nothing): Tap[tapT.T] = EmptyTap
+}
+
+object ZstdDictIO {
+  val DictionarySizeMbWarningThreshold: Int = 10
+
+  final case class WriteParam(
+    zstdDictSizeBytes: Int = WriteParam.DefaultZstdDictSizeBytes,
+    numElementsForSizeEstimation: Long = WriteParam.DefaultNumElementsForSizeEstimation,
+    trainingBytesTarget: Int = WriteParam.DefaultTrainingBytesTarget
+  )
+
+  object WriteParam {
+    val DefaultZstdDictSizeBytes: Int = 110 * 1024
+    val DefaultNumElementsForSizeEstimation: Long = 100L
+    val DefaultTrainingBytesTarget: Int = null.asInstanceOf[Int]
```

Comment:
This cast gives zero. Tried in `scala-cli`:
```
scala> val i: Int = null.asInstanceOf[Int]
val i: Int = 0
```
For primitive you MUST use option or box in `java.lang.Integer`

Reply:
Yikes! Moved to an Option